src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,914 @@
+    }
+    case 'UnaryExpression': {
+      const argType = typeCheckAndReturnType(node.argument, context, env)
+      const operator = node.operator
+      switch (operator) {
+        case '-':
+          // Only number/any type allowed
+          checkForTypeMismatch(node, argType, tNumber, context)
+          return tNumber
+        case '!':
+          // Only boolean/any type allowed
+          checkForTypeMismatch(node, argType, tBool, context)
+          return tBool
+        case 'typeof':
+          // No checking needed, typeof operation can be used on any type
+          return tString
+        default:
+          throw new TypecheckError(node, 'Unknown operator')
+      }
+    }
+    case 'BinaryExpression': {
+      return typeCheckAndReturnBinaryExpressionType(node, context, env)
+    }
+    case 'LogicalExpression': {
+      // Left type must be boolean/any
+      const leftType = typeCheckAndReturnType(node.left, context, env)
+      checkForTypeMismatch(node, leftType, tBool, context)
+
+      // Return type is union of boolean and right type
+      const rightType = typeCheckAndReturnType(node.right, context, env)
+      return mergeTypes(tBool, rightType)
+    }
+    case 'ArrowFunctionExpression': {
+      return typeCheckAndReturnArrowFunctionType(node, context, env)
+    }
+    case 'FunctionDeclaration':
+      // Only identifiers are used as function params in Source
+      const params = node.params as tsEs.Identifier[]
+      const expectedReturnType = getTypeAnnotationType(node.returnType, context, env)
+
+      const types = getParamTypes(params, context, env)
+      // Return type will always be last item in types array
+      types.push(expectedReturnType)
+      const fnType = tFunc(...types)
+
+      // Type check function body, creating new environment to store arg types, return type and function type
+      pushEnv(env)
+      params.forEach(param => {
+        setType(param.name, getTypeAnnotationType(param.typeAnnotation, context, env), env)
+      })
+      // Set unique identifier so that typechecking can be carried out for return statements
+      setType(RETURN_TYPE_IDENTIFIER, expectedReturnType, env)
+      setType(node.id!.name, fnType, env)
+      const actualReturnType = typeCheckAndReturnType(node.body, context, env)
+      env.pop()
+
+      if (
+        isEqual(actualReturnType, tVoid) &&
+        !isEqual(expectedReturnType, tAny) &&
+        !isEqual(expectedReturnType, tVoid)
+      ) {
+        // Type error where function does not return anything when it should
+        context.errors.push(new FunctionShouldHaveReturnValueError(node))
+      } else {
+        checkForTypeMismatch(node, actualReturnType, expectedReturnType, context)
+      }
+
+      // No need to save variable type again, return undefined type
+      return tUndef
+    case 'VariableDeclaration': {
+      const id = node.declarations[0].id as tsEs.Identifier
+      const init = node.declarations[0].init!
```

Comment:
Would be good to add a comment explaining why this non-null assertion is safe. Alternatively, you can add a check to narrow the type, throwing an error if it's `null` or `undefined`.